src/disco/containers/InstallButton.js

```diff
@@ -29,12 +32,48 @@ export class InstallButton extends React.Component {
     downloadProgress: 0,
   }
 
+  componentDidMount() {
+    const { dispatch, guid, installURL, slug } = this.props;
+    const payload = {slug, guid, url: installURL};
+    this.addonManager = new AddonManager(guid, installURL, this.statusChanged);
+    this.addonManager.getAddon()
+      .then(
+        () => dispatch({type: 'INSTALL_STATE', payload: {...payload, status: INSTALLED}}),
+        () => dispatch({type: 'INSTALL_STATE', payload: {...payload, status: UNINSTALLED}}));
+  }
+
+  statusChanged = (addonInstall) => {
+    const { dispatch, slug } = this.props;
+    if (addonInstall.state === 'STATE_DOWNLOADING') {
+      const downloadProgress = parseInt(
+        100 * addonInstall.progress / addonInstall.maxProgress, 10);
+      dispatch({type: 'DOWNLOAD_PROGRESS', payload: {slug, downloadProgress}});
+    } else if (addonInstall.state === 'STATE_INSTALLING') {
+      dispatch({type: 'START_INSTALL', payload: {slug}});
+    } else if (addonInstall.state === 'STATE_INSTALLED') {
+      dispatch({type: 'INSTALL_COMPLETE', payload: {slug}});
+    }
+  }
+
+  install() {
+    const { dispatch, slug } = this.props;
+    dispatch({type: 'START_DOWNLOAD', payload: {slug}});
+    this.addonManager.install();
+  }
+
+  uninstall() {
+    const { dispatch, slug } = this.props;
+    dispatch({type: 'START_UNINSTALL', payload: {slug}});
+    this.addonManager.uninstall()
```

Comment:
I think these functions that use dispatch should be getting passed in like they were before. That way we can test the component and just check "it calls the function" and to test these functions we can pull them off of `mapDispatchToProps` and verify they work as expected.